### ISSUE TITLE
SocketOption 128 resolves to unknown constant

### ISSUE THREAD

**Issue by etehtsea** [jnr/jnr-constants#16]
OS:
`15.6.0 Darwin Kernel Version 15.6.0: Thu Jun 23 18:25:34 PDT 2016; root:xnu-3248.60.10~1/RELEASE_X86_64 x86_64`

`jruby 9.1.3.0-SNAPSHOT (2.3.0) 2016-07-28 75b925c Java HotSpot(TM) 64-Bit Server VM 25.102-b14 on 1.8.0_102-b14 [darwin-x86_64]`

```ruby
pry(main)> s.getsockopt(Socket::SOL_SOCKET, Socket::SO_LINGER)
Errno::ENOPROTOOPT: Protocol not available - Protocol not available
from org/jruby/ext/socket/RubyBasicSocket.java:265:in `getsockopt'

pry(main)> Socket::SO_LINGER
=> 128
pry(main)> Socket::SOMAXCONN
=> 128
```

`getsockopt` method:

```ruby
    public IRubyObject getsockopt(ThreadContext context, IRubyObject _level, IRubyObject _opt) {
        Ruby runtime = context.runtime;

        SocketLevel level = SocketUtils.levelFromArg(_level);
        SocketOption opt = SocketUtils.optionFromArg(_opt);

        try {
            Channel channel = getOpenChannel();

            switch(level) {

            case SOL_SOCKET:
            case SOL_IP:
            case SOL_TCP:
            case SOL_UDP:

                if (opt == SocketOption.__UNKNOWN_CONSTANT__) {
                    throw runtime.newErrnoENOPROTOOPTError();
                }
```

`SocketOption.valueOf(128)` returns `__UNKNOWN_CONSTANT__`.

On MRI 2.3.1:

```ruby
[1] pry(main)> require 'socket'; s = Socket.tcp('localhost', 6379); s.getsockopt(Socket::SOL_SOCKET, Socket::SO_LINGER)
=> #<Socket::Option: INET SOCKET LINGER off 0sec>
[2] pry(main)> Socket::SO_LINGER
=> 128
[3] pry(main)> Socket::SOMAXCONN
=> 128
```

**Comment by etehtsea** [comment]
Place where `SO_LINGER` gets overriden by `SOMAXCONN`:

https://github.com/jnr/jnr-constants/blob/master/src/main/java/jnr/constants/ConstantSet.java#L117-L123

**Comment by etehtsea** [comment]
It's also defined in https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/ext/socket/RubySocket.java#L103

**Comment by headius** [comment]
Full backtrace for the error:

```
$ jruby -Xbacktrace.style=raw -rsocket -e "TCPSocket.new('google.com', 80).getsockopt(Socket::SOL_SOCKET, Socket::SO_LINGER)"
Errno::ENOPROTOOPT: Protocol not available - Protocol not available
             getStackTrace at java/lang/Thread.java:1589
          getBacktraceData at org/jruby/runtime/backtrace/TraceType.java:246
              getBacktrace at org/jruby/runtime/backtrace/TraceType.java:47
          prepareBacktrace at org/jruby/RubyException.java:235
                  preRaise at org/jruby/exceptions/RaiseException.java:214
                  preRaise at org/jruby/exceptions/RaiseException.java:181
                    <init> at org/jruby/exceptions/RaiseException.java:111
         newRaiseException at org/jruby/Ruby.java:4203
  newErrnoENOPROTOOPTError at org/jruby/Ruby.java:3616
                getsockopt at org/jruby/ext/socket/RubyBasicSocket.java:265
                      call at org/jruby/ext/socket/RubyBasicSocket$INVOKER$i$2$0$getsockopt.gen:-1
              cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:358
                      call at org/jruby/runtime/callsite/CachingCallSite.java:195
   invokeOther3:getsockopt at -e:1
                    <main> at -e:1
       invokeWithArguments at java/lang/invoke/MethodHandle.java:599
                      load at org/jruby/ir/Compiler.java:111
                 runScript at org/jruby/Ruby.java:834
               runNormally at org/jruby/Ruby.java:749
               runNormally at org/jruby/Ruby.java:767
               runFromMain at org/jruby/Ruby.java:580
             doRunFromMain at org/jruby/Main.java:425
               internalRun at org/jruby/Main.java:313
                       run at org/jruby/Main.java:242
                      main at org/jruby/Main.java:204
```

**Comment by headius** [comment]
@etehtsea Seems like this may be related to some other issues in JRuby. I mentioned jruby/jruby#3883 in #17 and you filed jruby/jruby#4040 recently. And of course there's jruby/jruby#4074.

I guess I just want to make sure we don't miss any open issues fixed by your work. We really appreciate your help!

**Comment by headius** [comment]
Fixed by #17.

**Comment by etehtsea** [comment]
I checked and all three are unrelated to this one.
